core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java

```diff
@@ -60,98 +61,104 @@
-    if (transformed) {
-      responseBuilder.message(transformBack(message));
-    } else {
-      responseBuilder.message(message);
-    }
-    if (previousCounterVar != null) {
-      responseBuilder.addVariable(counterVariableName, previousCounterVar);
-    } else {
-      responseBuilder.removeVariable(counterVariableName);
-    }
-    if (previousRootMessageVar != null) {
-      responseBuilder.addVariable(parentMessageProp, previousRootMessageVar);
-    } else {
-      responseBuilder.removeVariable(parentMessageProp);
-    }
-    return responseBuilder.build();
+    return processToApply(event, this);
   }
 
-  protected InternalEvent doProcess(InternalEvent event) throws MuleException {
-    try {
-      // TODO MULE-13052 Migrate Splitter and Foreach implementation to non-blocking
-      return Mono.just(event)
-          .then(request -> Mono
-              .from(processWithChildContext(request, ownedMessageProcessor, ofNullable(getLocation()))))
-          .onErrorMap(MessagingException.class, e -> {
-            if (splitter.equals(e.getFailingComponent())) {
-              // Make sure the context information for the exception is relative to the ForEach.
-              e.getInfo().remove(INFO_LOCATION_KEY);
-              return new MessagingException(event, e.getCause(), this);
+  @Override
+  public Publisher<InternalEvent> apply(Publisher<InternalEvent> publisher) {
+    return from(publisher).flatMap(originalEvent -> {
+
+      // Keep reference to existing rootMessage/count variables in order to restore later to support foreach nesting.
+      final String parentMessageProp = rootMessageVariableName != null ? rootMessageVariableName : ROOT_MESSAGE_PROPERTY;
+      final Object previousCounterVar = originalEvent.getVariables().containsKey(counterVariableName)
+          ? originalEvent.getVariables().get(counterVariableName).getValue()
+          : null;
+      final Object previousRootMessageVar =
+          originalEvent.getVariables().containsKey(parentMessageProp)
+              ? originalEvent.getVariables().get(parentMessageProp).getValue()
+              : null;
+
+      final InternalEvent requestEvent =
+          builder(originalEvent).addVariable(parentMessageProp, originalEvent.getMessage()).build();
+      return splitAndProcess(requestEvent)
+          .map(result -> {
+            final Builder responseBuilder = builder(result).message(originalEvent.getMessage());
+
+            // Restore original rootMessage/count variables.
+            if (previousCounterVar != null) {
+              responseBuilder.addVariable(counterVariableName, previousCounterVar);
             } else {
-              return e;
+              responseBuilder.removeVariable(counterVariableName);
             }
-          }).block();
-    } catch (Throwable throwable) {
-      throw rxExceptionToMuleException(throwable);
-    }
+            if (previousRootMessageVar != null) {
```

Comment:
Good point.  If an error with splitting occurs then this is not an issue but if error processing a part occurrs then need to take this into account, even if it was ignored in 3.x.